### ISSUE TITLE
Is there any way to move windows with super+button1+mouse drag

### ISSUE THREAD

**Issue by Notevill** [JLErvin/berry#60]
I want to manage windows like it's in 2bwm. `super+batton1+mouse drag` for moving windows and `super+batton3+mouse drag` for resizing. And this working by pressing on window itself not only on decorations.
If I right understand `bspwm` has `pointer` command for mouse grabbing. Is there a way to do the same with berry?

**Comment by JLErvin** [comment]
This isn't implemented right now, but this is something that I can add.

**Comment by Notevill** [comment]
I will waiting this feature.

**Comment by lucat1** [comment]
I'd love this feature too! Isn't this possible with a tool such as `sxhkd`?

**Comment by JLErvin** [comment]
@lucat1 As far as I know, no. Motion events that spawn processes disabled in `sxhkd` after `0.5.5` https://github.com/baskerville/sxhkd/commit/07595089fee6aac5953af8acb2f2e98682fa03d5. `berry` handles motion events internally as a result, so this is something I would have to implement on the backend. Finals are done next week and I should have some time to implement this.

**Comment by mintyleaf** [comment]
like a workaround i maked this script
```
#!/bin/sh
eval $(xdotool getactivewindow getwindowgeometry --shell)
eval $(xdotool getmouselocation --shell)
berryc window_move_absolute $(echo "$X - $WIDTH / 2" | bc) $(echo "$Y - $HEIGHT / 2" | bc)
berryc window_move 0 0
```
that comes in sxhkd like this
```
super + button1$
    ~/.scripts/brymmove.sh$
```

**Comment by bcd0101** [comment]
@mintyleaf very useful your script. Thanks for sharing.

Would be awesome if we have a way to resize window using another script like that.

**Comment by mintyleaf** [comment]
> @mintyleaf very useful your script. Thanks for sharing.
> 
> Would be awesome if we have a way to resize window using another script like that.

 i think this thing can be achivied through window_resize_absolute and some calculactions between window pos from xdotool with bc

**Comment by Notevill** [comment]
Make this script for resize

```
while [ ! $1 ]; do
	eval $(xdotool getactivewindow getwindowgeometry --shell)
	WX=$X
	WY=$Y
	eval $(xdotool getmouselocation --shell)
	berryc window_resize_absolute $(echo "$X - $WX" | bc) $(echo "$Y - $WY" | bc)
done

pkill -o -f $0
```

In sxhkdrc:
for start resizing 

```
super + button3
	berryc pointer_focus && berry_resize
```

for stop resizing

```
@button3
	berry_resize stop
```

This allow me resize window while I hold and drug button3. 
And same I do for moving window.

**Comment by mintyleaf** [comment]
@Notevill 
there is little typo that can confuse someone (me too))
for stop resizing:
```
@super + button3
    berry_resize stop

**Comment by mintyleaf** [comment]
i rewrote theese scripts for more intuitive and "draggable" behaviour
it's using not absolute moving/resizing, that which allow to respect global gaps
also it prevents the jumping to the mouse pos, using relative positioning

sxhkd
```
super + button1
    ~/.scripts/brymove.sh

@super + button1
    ~/.scripts/brymove.sh stop

super + button3
   ~/.scripts/brysize.sh

@super + button3
   ~/.scripts/brysize.sh stop
```

brymove.sh
```
#!/bin/sh

eval $(xdotool getmouselocation --shell)
SX=$X
SY=$Y

while [ ! $1 ]; do
    eval $(xdotool getmouselocation --shell)
    berryc window_move $(echo "$X - $SX" | bc) $(echo "$Y - $SY" | bc)
    SX=$X
    SY=$Y
done

pkill -f $0
```

brysize,sh
```
#!/bin/sh

eval $(xdotool getmouselocation --shell)
SX=$X
SY=$Y

while [ ! $1 ]; do
    eval $(xdotool getmouselocation --shell)
    berryc window_resize $(echo "$X - $SX" | bc) $(echo "$Y - $SY" | bc)
    SX=$X
    SY=$Y
done

pkill -f $0
```

**Comment by Notevill** [comment]
@mintyleaf  You add this lines too, so it's not typo?

```
@button1
    ~/.scripts/brymove.sh stop

@button3
    ~/.scripts/brysize.sh stop
```
I also have bug when moving or resizing didn't stops after I release buttons.

**Comment by mintyleaf** [comment]
> @mintyleaf You add this lines too, so it's not typo?
> 
> ```
> @button1
>     ~/.scripts/brymove.sh stop
> 
> @button3
>     ~/.scripts/brysize.sh stop
> ```
> 
> I also have bug when moving or resizing didn't stops after I release buttons.

yup
this is not a typo, sorry)
for proper working script needs to be stopped after super + button release and simple button release too

**Comment by Notevill** [comment]
>
> for proper working script needs to be stopped after super + button release and simple button release too

No I have this bug with `@button1` too. I think it's sxhkd issue, because multiple button1 click with or without super not stops window moving.

**Comment by bcd0101** [comment]
@Notevill I'm using the scripts without the described bug and without calling additional stop after release super + button1 or button 2

**Comment by mintyleaf** [comment]
hey, seems like i'm faced sxhkd bug too
when config has @button3 event all working fine for script, but in any other windows rmb didn't captured at all)
but @button1 event didn't affect for other windows like b3

so, now my sxhkdrc looks like this (i replaced two b1 events with any + @b1):
```
 super + button1
     ~/.scripts/brymove.sh
  any + @button1
     ~/.scripts/brymove.sh stop
  super + button3
    ~/.scripts/brysize.sh
 super + @button3
    ~/.scripts/brysize.sh stop
```

**Comment by TheM4ze** [comment]
I put both scripts in my ~/.bin then gave permission and chmod +x both of them.
Copied this into sxhkd [ ofc changed to my path, ~/.bin/bla bla bla/brymove.sh ]

> super + button1
     ~/.scripts/brymove.sh
  any + @button1
     ~/.scripts/brymove.sh stop
  super + button3
    ~/.scripts/brysize.sh
 super + @button3
    ~/.scripts/brysize.sh stop <

Restart berry and sxhkd but mine acts weird. 

Instead of resizing windows+moving it messes up my [ Return,Erase,Shift and ctrl ] keys. Don’t know why tbh,when I remove them it works as should. Can’t move any window at all,have to use [ super + tab } to focus on windows.
Am I missing something?

**Comment by Notevill** [comment]
Are we still waiting for a native solution from berry?

**Comment by mintyleaf** [comment]
> Are we still waiting for a native solution from berry?

ofc, cause using scripts like this for "drag'n'droppy" behaviour is cpu heavy and sxhkd is very buggy with mouse button events

**Comment by JLErvin** [comment]
I'm currently working on native move/resize support with mod keys. Development overall has been pretty slow given the pandemic - but I am working on it. I'll keep you updated but hopefully it should be done in a few days.

**Comment by TheM4ze** [comment]
> > Are we still waiting for a native solution from berry?
> 
> ofc, cause using scripts like this for "drag'n'droppy" behaviour is cpu heavy and sxhkd is very buggy with mouse button events

Yep sxhkd is buggy with mouse buttons. Today I've tried using the 2 scripts but couldn't get it to work at all. Guess I'll have to wait :(

**Comment by JLErvin** [comment]
> > > Are we still waiting for a native solution from berry?
> > 
> > 
> > ofc, cause using scripts like this for "drag'n'droppy" behaviour is cpu heavy and sxhkd is very buggy with mouse button events
> 
> Yep sxhkd is buggy with mouse buttons. Today I've tried using the 2 scripts but couldn't get it to work at all. Guess I'll have to wait :(

Hang in there - it's coming soon :)

**Comment by TheM4ze** [comment]
> > > > Are we still waiting for a native solution from berry?
> > > 
> > > 
> > > ofc, cause using scripts like this for "drag'n'droppy" behaviour is cpu heavy and sxhkd is very buggy with mouse button events
> > 
> > 
> > Yep sxhkd is buggy with mouse buttons. Today I've tried using the 2 scripts but couldn't get it to work at all. Guess I'll have to wait :(
> 
> Hang in there - it's coming soon :)

No worries I can live a bit longer without it (notice A BIT longer) :P

**Comment by JLErvin** [comment]
Native move/resizing support is now available on the development branch, `huckleberry`. You can preview/use it [here](https://github.com/JLErvin/berry/tree/huckleberry).

From a user perspective, moving windows defaults to `Mod4` (Usually Windows keys) and resizing windows defaults to `Mod1` (Usually Alt). You can change this using the `berryc` commands `move_mask` and `resize_mask`, respectively. They can take on any value of `Mod1` through `Mod5`. If both of them take on the same value, `move_mask` takes precedence. 

Give it a try and let me know if there are any issues.

**Comment by TheM4ze** [comment]
> Native move/resizing support is now available on the development branch, `huckleberry`. You can preview/use it [here](https://github.com/JLErvin/berry/tree/huckleberry).
> 
> From a user perspective, moving windows defaults to `Mod4` (Usually Windows keys) and resizing windows defaults to `Mod1` (Usually Alt). You can change this using the `berryc` commands `move_mask` and `resize_mask`, respectively. They can take on any value of `Mod1` through `Mod5`. If both of them take on the same value, `move_mask` takes precedence.
> 
> Give it a try and let me know if there are any issues.

Was just playing around and had absolutely no issues!

super + ~button1 to move
alt + ~button1 to resize

**Comment by Harjuu** [comment]
I found the resizing to be pretty slow compared to other WMs. When resizing too fast I get a very noticable freeze that I don't get with for example spectrwm or sowm. 

Other than that, fantastic!

**Comment by TheM4ze** [comment]
> I found the resizing to be pretty slow compared to other WMs. When resizing too fast I get a very noticable freeze that I don't get with for example spectrwm or sowm.
> 
> Other than that, fantastic!
> 
> Edit: Actually, it seems to be only when increasing the size of the window, shrinking works flawlessly

Hmmmm just tried out both shrinking and increasing size and both work flawlessly for me. Couldn't notice any difference compared to other wm's.regarding speed or really anything.

**Comment by JLErvin** [comment]
@Harjuu are you using a compositor?

**Comment by Notevill** [comment]
There is no way to specify `super+button1` to move and `super+button3` to resize, only `<mod>+button1`, isn't it?

**Comment by JLErvin** [comment]
Right now, no. But this can be easily implemented.

**Comment by Notevill** [comment]
@Harjuu, did you mean freezes like this?
![screencast](https://user-images.githubusercontent.com/34532252/81043735-47f60700-8ecc-11ea-9c44-923d19efd696.gif)

**Comment by Harjuu** [comment]
Hmm strange. @JLErvin yes I'm using picom, but the only difference when I kill picom and try is that I get massive screen tearing instead from the window not keeping up.

I don't think it's my pc as I'm not having any problems with this on any other wm, with or without compositor, and I don't have a slow pc by any means.

@Notevill I think that comes very close, difference being I'm not using any borders or titlebars, so there's nothing in my window that always "keeps up" with the mouse like it does in your case, which makes it look and feel a lot worse. Mine stutters a lot more though, I'll see if I can record it

**Comment by Notevill** [comment]
It's not working when NumLock on.

**Comment by Harjuu** [comment]
![2020-05-05-11-34-32](https://user-images.githubusercontent.com/16627772/81053098-dc5c6b80-8ec4-11ea-9dc4-883f336bee14.gif)

First few times is with picom, I then kill it and try to resize and the window actually crashes after I resize it two times. Never saw that before. Also noticed that on the video it actually looks the same with/without picom, but on my screen the window slowly catches up to the cursor when not using picom, which doesn't get picked up in the recordning.

**Comment by JLErvin** [comment]
@Harjuu can you try on the lastest version of the development branch and see if it's any different?

**Comment by Harjuu** [comment]
No difference unfortunately. 

![2020-05-07-14-26-27](https://user-images.githubusercontent.com/16627772/81294536-56305880-906f-11ea-8fa7-273f2f1a31a0.gif)

Let me know if you think there could be something on my end affecting this

**Comment by JLErvin** [comment]
Well, it took me forever, but the input lag resizing is finally fixed. If you are experiencing input lag while resizing, try the following:
```
berryc pointer_interval 17
```
This sets a minimum amount of time between two pointer motion events, which prevents events from accumulating faster than we can resize windows.

Moving and resizing windows using super+button+drag is included as of 0.1.6.

See the man pages for appropriate move/resize mask values.

Closing

**Comment by arcolinuxz** [comment]
Is this now implemented into Berry or do we need to use the scripts?
